cdap-app-fabric/src/main/java/co/cask/cdap/app/DefaultAppConfigurer.java

```diff
@@ -171,12 +179,25 @@ public void addWorkflow(Workflow workflow) {
   @Override
   public void addService(Service service) {
     Preconditions.checkArgument(service != null, "Service cannot be null.");
+    // check that a system service is only used in system namespace
+    if (!deployNamespace.equals(Id.Namespace.fromEntityId(NamespaceId.SYSTEM))) {
+      TypeToken<?> type = TypeToken.of(service.getClass()).resolveType(Service.class.getTypeParameters()[0]);
+      if (SystemServiceConfigurer.class.isAssignableFrom(type.getRawType())) {
+        throw new IllegalArgumentException(String.format(
+          "Invalid service '%s'. Services can only use a SystemServiceConfigurer if the application is "
+            + "deployed in the system namespace.", service.getClass().getSimpleName()));
+      }
+    }
+
+    DefaultSystemTableConfigurer systemTableConfigurer = new DefaultSystemTableConfigurer();
     DefaultServiceConfigurer configurer = new DefaultServiceConfigurer(service, deployNamespace, artifactId,
-                                                                       artifactRepository, pluginInstantiator);
+                                                                       artifactRepository, pluginInstantiator,
+                                                                       systemTableConfigurer);
```

Comment:
I think creating them here is in some ways analogous to the system creating them up-front, as this is how we would avoid needing getOrCreate() whenever the table is used at runtime.

That being said, one way I can think of to separate them is to make system table creation part of the bootstrap process by adding a new type of "create system table" bootstrap step.  Though if we do it that way it may seem kind of weird to have some tables created that way and others created in the code.

Reply:
We decided to keep this as-is today, but we may separate it out in the future